src/V3Sched.cpp

```diff
@@ -182,7 +192,11 @@ LogicClasses gatherLogicClasses(AstNetlist* netlistp) {
             } else if (senTreep->hasCombo()) {
                 UASSERT_OBJ(!senTreep->sensesp()->nextp(), activep,
                             "combinational logic with additional sensitivities");
-                result.m_comb.emplace_back(scopep, activep);
+                if (onlyHasPost(activep)) {
+                    result.m_post.emplace_back(scopep, activep);
+                } else {
+                    result.m_comb.emplace_back(scopep, activep);
+                }
```

Comment:
Yep, I was trying out two variants and left both in accidentally.

I think `AstAssignPre` is put in the same active as `Post`, so I'll have to split that up in `V3Delayed`.

Reply:
There's also a problem with `V3SchedReplicate`, which expects an acyclic graph, otherwise it recurses forever. But from what I understand post logic should be ran through it.

Reply:
I don't think post logic needs to go through that. Acyclic is only for combinational logic, which we didn't use to have in post until now.

Reply:
I'm starting to wonder if we should reconsider making a proper sentree for these Post nodes, rather than making them Comb. Given there is currently only one sen tree per suspendable process, we could just attach this in the timing pass to the process then grab it in V3Delay?

Reply:
> I don't think post logic needs to go through that. Acyclic is only for combinational logic, which we didn't use to have in post until now.

Skipping `Acyclic` is not a problem, skipping `Replicate` is. If I put post logic into `nba` after `partition` everything is fine, but if I wait until after `replicate`, 30 regression tests fail. I think it's because of combo logic dependant on it. Guessing it gets deleted because the upstream logic seemingly doesn't exist.

> Given there is currently only one sen tree per suspendable process

Not sure what you mean? There's multiple sentrees per suspendable process – as many as there are event controls in it plus possibly one for delays.

Reply:
> Not sure what you mean? There's multiple sentrees per suspendable process – as many as there are event controls in it plus possibly one for delays.

What I meant here is that all of them get combined into a single sentree, but I realized that is not the right way of handling it as it can lead to a large number of extra sentrees.

Reply:
I tried multiple ways of handling `post` logic separately, but I always ended up with special case handling for those `dlyvset` variables in `V3SchedReplicate`.

I propose a different solution:
1. For each `AlwaysPost` triggered by timing logic, keep a pointer to that upstream logic.
2. Then, in `V3Timing`, clone that `AlwaysPost` into actives corresponding to each sentree from the upstream logic. That way it'll be treated as clocked, and there's no cycle problem.

That's potentially a lot of clones, but I think this solution is simple, clean, and fits in nicely with how scheduling works. In the future the number of clones will be reduced once we start properly analyzing what sentrees affect which writes.

What do you think?

Reply:
Let me try to expand on this problem, so hopefully we can find a solution.

Classically, before timing support, all `Post` logic had a single SenTree: the SenTree of the always that triggered it:

```
always @(posedge foo) a <= b;
```

Got turned into (with arbitrary trigger indices):
```
// In eval_triggers_act:
actTrigVec.at(2) = foo & ~prev_foo;

// (act triggers accumulated into nba triggers in the eval loop)
nbaTrigVec |= actTrigVec;

// in eval_nba
if (nbaTrigVec.at(2)) <PostLogic>
```

Because of delays, we can have multiple senTrees triggering a Post corresponding to the NBAs in the process:

```
initial forever begin
  if (c) @(posedge foo); else @posedge(bar);
  a <= b;
end
```

The ideal situation for this would be to render it as:

```
// In eval_triggers_act:
actTrigVec.at(2) = foo & ~prev_foo;
actTrigVec.at(3) = bar & ~prev_bar;

// (act triggers accumulated into nba triggers in the eval loop)
nbaTrigVec |= actTrigVec;

// in eval_nba
if (nbaTrigVec.at(2) || nbaTrigVec.at(3)) <PostLogic>
```

I think this is what we were trying to achieve by making the post logic comb. The crux of the problem is that the current scheduling and ordering code does not handle clocked logic with multiple SenTrees (even though multiple SenTrees are semantically equivalent to a single SenTree with their combines SenItems).

---

> Then, in V3Timing, clone that AlwaysPost into actives corresponding to each sentree from the upstream logic.

Let's try hard not to clone code.

What I suggest instead, is to put the AlwaysPost under an Active with a SenTree that is the combination of all involved SenTrees. See `combineDomains` in V3Order.cpp on how to combine SenTrees, most notably, if doing it after V3ActiveTop, then you need to add/uniqueify the resulting SenTree with the ones in the Netlist using SenTreeFinder. 

Implementation note: Whether you need/want to hold onto the upstream logic to do this is up to you, I can't see it clearly. Ideally we would create the AstPost under the correct Active already in V3Delayed, we should try strongly to achieve this to keep the passes less interdependent, and it's still better to carry data/protocol from V3Timing to V3Delayed than from V3Timing and V3Delayed both to V3Sched, so use your judgement.

The effect of this is that the output will be:
```
// In eval_triggers_act:
actTrigVec.at(2) = foo & ~prefv_foo;
actTrigVec.at(3) = bar & ~prefv_bar;
actTrigVec.at(4) = foo & ~prev_foo || bar & ~prev_bar; // This is just actTrigVec.at(2) || actTrigVec.at(3), but stored back in the trigger vector.

// (act triggers acumulated into nba triggers in the eval loop)
nbaTrigVec |= actTrigVec;

// in eval_nba
if (nbaTrigVec.at(4)) <PostLogic>
```

What we are losing here, is one byte of memory for each unique combination of SenTrees we need to create. (Plus processing those but that should be cheap in the grand scheme of things on large designs). With a lot of wacky event control combinations that might add up, but should work, but in exchange will not need to duplicate code (and output code size is the number one Verilator performance problem).

I believe this will get you to a working state.

---

Once we done that, I suggest in a later patch, we will fix the original deficiency in the current scheduling/ordering implementation, namely, as I said it before: that the current scheduling and ordering code does not handle clocked logic with multiple SenTrees. We should be able to improve on this by making AstActive hold a list of SenTrees, rather than just one, but this needs updating all bits involved, including at least V3Sched and V3Order, and probably some of the earlier parts, so probably better to do this as a separate patch to keep things manageable.

---

What are your thoughts?

Reply:
Implementation note: Can't you just defer adding the Post nodes to the tree until the end of the suspendable in V3Delayed? by then you should have been able to pick up all SenTrees (or wors case foreach through the suspendable process in V3Delayed), so you can create an Active with the combined SenTree?

Reply:
I think we're on the same page, it's just I thought merging the sentrees wouldn't be acceptable. But I can do that, sure.

I actually thought about actives holding multiple sentrees, but like you said, it's a pretty big change on its own.

> Implementation note: Can't you just defer adding the Post nodes to the tree until the end of the suspendable in V3Delayed? by then you should have been able to pick up all SenTrees (or wors case foreach through the suspendable process in V3Delayed), so you can create an Active with the combined SenTree?

Yes, makes sense. The foreach solution appeals to me because the gathering code could be shared between `V3Delayed` and `V3Timing`, and it'd be less confusing overall. Which one do you prefer?

Reply:
Using foreach we would effectively iterate the process twice would we not? If we can easily avoid that then best do it with one traversal (as part of the visitor)

Reply:
Ok, now `V3Delayed` builds an explicit sentree for `AstAlwaysPost`s.

There was still the problem with the `DelayScheduler` var not being present yet when `V3Delayed` is executed (so the `@ [true] delayScheduler.awaitingCurrentTime()` didn't exist at that time). So I moved the code around, and now all of `V3Timing` is done at once, and `AstCAwait`s store pointers to sentrees that trigger them (or, another interpretation is that they store external domains for variables written to after the await is done).

Then the `TimingKit` is created in `V3Sched directly` by `foreach`ing the relevant nodes.

Also, fork child statements are moved to separate nodes later, at the end of `V3Sched`. This is to simplify gathering timing sentrees (we don't need to traverse `CCall`s).

What do you think? (don't look at the code quality yet, there's still cleanup to do, like the two timing visitors have to be merged. It's a big diff, but like I said, it's mostly just rearranged code)

Reply:
See below, I like it. Please ping me when you are happy with it and want me to do another once-over.

Reply:
I'm pretty happy with it now.